core/src/main/java/io/grpc/internal/ServerImpl.java

```diff
@@ -320,24 +302,88 @@ public void run() {
         timeoutMicros,
         TimeUnit.MICROSECONDS);
     }
+  }
+
+  @VisibleForTesting
+  static class StreamCreatedCallback implements Runnable {
+    private final ServerStream stream;
+    private final String methodName;
+    private final Metadata headers;
+    private final JumpToApplicationThreadServerStreamListener jumpListener;
+    private final HandlerRegistry registry;
+    private final DecompressorRegistry decompressorRegistry;
+    private final Future<?> timeout;
+
+    StreamCreatedCallback(ServerStream stream, String methodName, Metadata headers,
+        JumpToApplicationThreadServerStreamListener jumpListener, HandlerRegistry registry,
+        DecompressorRegistry decompressorRegistry, Future<?> timeout) {
+      this.stream = stream;
+      this.methodName = methodName;
+      this.headers = headers;
+      this.jumpListener = jumpListener;
+      this.registry = registry;
+      this.timeout = timeout;
+      this.decompressorRegistry = decompressorRegistry;
+    }
+
+    @Override
+    public void run() {
+      ServerStreamListener listener = NOOP_LISTENER;
+      try {
+        ServerMethodDefinition<?, ?> method = registry.lookupMethod(methodName);
+        if (method == null) {
+          stream.close(
+              Status.UNIMPLEMENTED.withDescription("Method not found: " + methodName),
+              new Metadata());
+          timeout.cancel(true);
+          return;
+        }
+        listener = startCall(method);
+      } catch (Throwable t) {
+        stream.close(Status.fromThrowable(t), new Metadata());
+        timeout.cancel(true);
+        throw Throwables.propagate(t);
+      } finally {
+        jumpListener.setListener(listener);
+      }
+    }
 
     /** Never returns {@code null}. */
-    private <ReqT, RespT> ServerStreamListener startCall(ServerStream stream, String fullMethodName,
-        ServerMethodDefinition<ReqT, RespT> methodDef, Future<?> timeout,
-        Metadata headers) {
-      // TODO(ejona86): should we update fullMethodName to have the canonical path of the method?
+    private <ReqT, RespT> ServerStreamListener startCall(
+        ServerMethodDefinition<ReqT, RespT> methodDef) {
+      if (decompressorRegistry != null) {
+        stream.setDecompressionRegistry(decompressorRegistry);
+      }
+
+      String messageEncoding = headers.get(GrpcUtil.MESSAGE_ENCODING_KEY);
+      if (messageEncoding != null && stream instanceof AbstractStream) {
+        try {
+          // InProcessTransport doesn't support decompressors, so cast :(
+          ((AbstractStream<?>)stream).setDecompressor(messageEncoding);
```

Comment:
I agree with @zhangkun83 raising #1076, because it was an internal detail that had no need to be exposed. Like he said though, `ServerStream` may be a different case.

I still claim that if we are going to call `setDecompressor()` here, we should add the method to `Stream` or `ServerStream`. If `Stream` doesn't make sense, that's fine; we would add it to `ServerStream` then. I'm still looking at whether `setDecompressor()` makes more sense here or some place else.